feat(analysis/normed_space): more facts about operator norm

jlpaca opened the pull request:
refactor some of `bounded_linear_maps.lean` and `operator_norm.lean`, and a couple of additional results about bounded linear maps.

Identical to #919, except filing request from a branch of `leanprover-community/mathlib` so that the CI tests would work.

TO CONTRIBUTORS:

Make sure you have:

  * [x] reviewed and applied the coding style: [coding](https://github.com/leanprover/mathlib/blob/master/docs/style.md), [naming](https://github.com/leanprover/mathlib/blob/master/docs/naming.md)
  * [ ] for tactics:
     * [ ] added or adapted documentation in [tactics.md](https://github.com/leanprover/mathlib/blob/master/docs/tactics.md)
     * [ ] write an example of use of the new feature in [tactics.lean](https://github.com/leanprover/mathlib/blob/master/tests/tactics.lean)
  * [x] make sure definitions and lemmas are put in the right files
  * [x] make sure definitions and lemmas are not redundant

If this PR is related to a discussion on Zulip, please include a link in the discussion.

For reviewers: [code review check list](https://github.com/leanprover/mathlib/blob/master/docs/code-review.md)

jlpaca committed:
refactor(analysis/normed_space): refactor and additional lemmas

- rename `bounded_linear_maps` to `bounded_linear_map`, `operator_norm` to `op_norm`.

- refactor operator norm with an equivalent definition.

- change some names and notation to be more consistent with conventions elsewhere
  in mathlib: replace `bounded_by_*` with `le_*`, `operator_norm_homogeneous` with
  `op_norm_smul`.

- more simp lemmas for bounded_linear_map.

- additional results: lipschitz continuity of the operator norm, also
  that it is submultiplicative.

jlpaca committed:
chore(analysis/normed_space/operator_norm): add attribution

jcommelin commented on src/analysis/normed_space/operator_norm.lean:
```diff
@@ -1,193 +1,183 @@
 
-@[simp] lemma bounded_linear_maps.smul_coe {A : L(E, F)} {x : E} {c : k} :
-  (c • A) x = c • (A x) := rfl
+def to_linear_map : linear_map _ E F :=
```
Should there also be a coercion from bounded linear maps to linear maps, defined as `to_linear_map`?

jlpaca replied:
Added in a0d68f11f97256b340e23eada235975862c8350b .

jcommelin commented on src/analysis/normed_space/operator_norm.lean:
```diff
@@ -1,193 +1,183 @@
-variables [normed_field k] [normed_space k E] [normed_space k F]
-open lattice set
+/-- The operator norm of a bounded linear map is the inf of all its bounds. -/
+def op_norm := real.Inf { c | c ≥ 0 ∧ ∀ x, ∥f x∥ ≤ c * ∥x∥ }
```
I'm not up to speed on `real.Inf`. Can't you just write `Inf`, and have type class inference figure out that you mean `real.Inf`?
This might also apply in other cases.

jlpaca replied:
That works. Thanks for the catch.

Changes added in a270440f2e98cc5c077cd8daced7ffb3c0129315 .

jcommelin commented on src/analysis/normed_space/operator_norm.lean:
```diff
@@ -1,193 +1,183 @@
-  { refine (div_le_iff _).2 (H x),
-    simpa [ne.symm h] using le_iff_eq_or_lt.1 (norm_nonneg x) }
-end
+/-- This is the fundamental property of the operator norm: ∥f x∥ ≤ ∥f∥ * ∥x∥. -/
```
```suggestion
/-- The fundamental property of the operator norm: ∥f x∥ ≤ ∥f∥ * ∥x∥. -/
```

jcommelin commented on src/analysis/normed_space/operator_norm.lean:
```diff
@@ -1,193 +1,183 @@
+              (le_op_norm _ _) (op_norm_nonneg _)
+  end⟩)
+
+/-- bounded linear maps are lipschitz continuous. -/
```
```suggestion
/-- Bounded linear maps are lipschitz continuous. -/
```

jcommelin reviewed (commented):
Looks pretty good to me. Some minor comments.

jlpaca committed:
style(analysis/normed_space/operator_norm): use namespace `real`

- open `real` instead of `lattice` and omit spurious prefixes.

jlpaca committed:
feat(analysis/normed_space/operator_norm): coercion to linear_map

jlpaca committed:
style(analysis/normed_space/bounded_linear_maps): minor edits

- extract variables for brevity of theorem statements.
- more consistent naming of variables.

jlpaca committed:
feat(analysis/normed_space/operator_norm): add constructor of bounded_linear_map from is_bounded_linear_map

jlpaca committed:
Merge branch 'master' of https://github.com/leanprover-community/mathlib into refactor-op-norm

jlpaca committed:
fix(analysis/normed_space/operator_norm): remove spurious explicit argument

sgouezel commented on src/analysis/normed_space/operator_norm.lean:
```diff
@@ -1,193 +1,190 @@
--- introduce the local notation L(E,F) for the set of bounded linear maps.
-local notation `L(` E `,` F `)` := @bounded_linear_maps k E F _ _ _
+/-- Construct bounded linear map from is_bounded_linear_map -/
+def from_is_bounded_linear_map {f : E → F} (hf : is_bounded_linear_map k f) :
```
It would be better to call this `is_bounded_linear_map.to_bounded_linear_map` (outside of the `bounded_linear_map` namespace). In this way, if you have an assumption `h : is_bounded_linear_map f` somewhere, then `h.to_bounded_linear_map` gives direct access to the corresponding bounded linear map.

jlpaca replied:
effected in ef8e8e5c1338234f5511d9401d610a3dcde9e2af .

jlpaca committed:
fix(analysis/normed_space): type of bounded linear maps

- change the definition of bounded_linear_map to be a type rather than
  the corresponding subspace, and mark it for unfolding.

- rename `bounded_linear_map.from_is_bounded_linear_map` to `is_bounded_linear_map.to_bounded_linear_map`.

jlpaca committed:
feat(analysis/normed_space): analysis results for bounded_linear_maps

avigad reviewed (dismissed):
This looks very good, and I appreciate all the work you (@jlpaca and @sgouezel) have put into this! To avoid holding things up, I will approve it right away. But it would be helpful to have a summary of what is changing in why. Tracing back through the previous PR's and all the discussion, it is hard to follow the main threads. In particular, I would really appreciate a summary explanation of:

- why bundling bounded linear maps is a good thing, and what problems it solves
- what the negatives are, and what concerns you have (e.g. the multiple ways of stating that bounded linear maps are continuous)
- how it affects that design of future libraries that will depend on bounded linear maps (like `fderiv`), and what the appropriate usage is.
- what the `@[reducible] def bounded_linear_map` trick is doing, and why it is necessary.

Commit:
Merge branch 'master' into 'refactor-op-norm'

sgouezel committed:
Merge branch 'master' into refactor-op-norm

jcommelin committed:
Merge branch 'master' into refactor-op-norm

jlpaca committed:
Merge branch 'master' into refactor-op-norm
